services/device/geolocation/geoclue_client_object.h

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) 2023 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#ifndef BRAVE_SERVICES_DEVICE_GEOLOCATION_GEOCLUE_CLIENT_OBJECT_H_
+#define BRAVE_SERVICES_DEVICE_GEOLOCATION_GEOCLUE_CLIENT_OBJECT_H_
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "base/functional/callback_forward.h"
+#include "base/memory/scoped_refptr.h"
+#include "base/memory/weak_ptr.h"
+#include "dbus/bus.h"
+#include "dbus/message.h"
+#include "dbus/object_proxy.h"
+#include "dbus/property.h"
+
+namespace device {
+
+class GeoClueClientObject {
+ public:
+  static constexpr char kServiceName[] = "org.freedesktop.GeoClue2";
+  static constexpr char kManagerInterfaceName[] =
+      "org.freedesktop.GeoClue2.Manager";
+  static constexpr char kManagerObjectPath[] =
+      "/org/freedesktop/GeoClue2/Manager";
+  static constexpr char kInterfaceName[] = "org.freedesktop.GeoClue2.Client";
+  static constexpr char kLocationInterfaceName[] =
+      "org.freedesktop.GeoClue2.Location";
+
+  struct LocationProperties : public dbus::PropertySet {
+    dbus::Property<double> latitude;
+    dbus::Property<double> longitude;
+    dbus::Property<double> accuracy;
+    dbus::Property<double> altitude;
+    dbus::Property<double> speed;
+    dbus::Property<double> heading;
+
+    explicit LocationProperties(dbus::ObjectProxy* proxy);
+    ~LocationProperties() override;
+
+    using dbus::PropertySet::GetAll;
+    void GetAll(base::OnceCallback<void()> on_got_all);
+
+    // dbus::PropertySet:
+    void OnGetAll(dbus::Response* response) override;
+
+   private:
+    base::OnceCallback<void()> on_got_all_;
+  };
+
+  struct Properties : public dbus::PropertySet {
+    dbus::Property<std::string> desktop_id;
+    dbus::Property<uint32_t> requested_accuracy_level;
+
+    explicit Properties(dbus::ObjectProxy* proxy);
+    ~Properties() override;
+  };
+
+  using LocationChangedCallback =
+      base::RepeatingCallback<void(std::unique_ptr<LocationProperties>)>;
+  using ErrorCallback = base::RepeatingCallback<void(const std::string&)>;
+
+  struct CreateParams {
+    CreateParams();
+    ~CreateParams();
+    CreateParams(const CreateParams&);
+    CreateParams& operator=(const CreateParams&);
```

Comment:
nit: for this to properly `std::move` you also need to declare rvalue assignment operator and constructor (preferrable with `noexcept`). that's the sad reality of c++ - when you declare a copy constructor/assignment operator, then you effectively disable default move operations :/

Reply:
Cool! Thanks for explaining that :smile: